BParkingNano/plugins/ElectronMerger.cc

```diff
@@ -38,93 +50,134 @@ class ElectronMerger : public edm::global::EDProducer<> {
-
-  std::vector<int> alreadySavedLPT;
-  alreadySavedLPT.resize(lowpt->size(), 0);
-
-  for(auto muonTrg : *trgMuon) {
-
-    int icount = -1;
-    for(auto ele : *pf) {
-      ++icount;
-      if(alreadySavedPF[icount]) continue;
-
-      if((reco::deltaR(ele, muonTrg) < drTrg_cleaning_ && drTrg_cleaning_ != -1) ||
-	 (std::fabs(ele.vz() - muonTrg.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ != -1)) continue;
-
-      ele.addUserInt("isPF", 1);
-      ele.addUserInt("isLowPt", 0);
-      ele.addUserFloat("ptBiased", 20.);
-      ele.addUserFloat("unBiased", 20.);
-      ele.addUserFloat("mvaId", 20.);
-      ele.addUserFloat("chargeMode", ele.charge());
-      alreadySavedPF[icount] = 1;
-      out->push_back(ele);
-    }
+  edm::Handle<edm::ValueMap<float> > ptBiased;
+  evt.getByToken(ptBiased_src_, ptBiased);
+  edm::Handle<edm::ValueMap<float> > unBiased;
+  evt.getByToken(unBiased_src_, unBiased);
+  edm::Handle<edm::ValueMap<float> > mvaId;  
+  evt.getByToken(mvaId_src_, mvaId);
 
-    icount = -1;
-    for(auto ele : *lowpt) {
-      ++icount;
-      if(alreadySavedLPT[icount]) continue;
-
-      if((reco::deltaR(ele, muonTrg) < drTrg_cleaning_ && drTrg_cleaning_ != -1) ||
-	 (std::fabs(ele.vz() - muonTrg.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ != -1)) continue;
-
-      bool clean_out = false;
-      for(const auto& pfele : *pf) {
-
-	clean_out |= (
-		      fabs(pfele.vz() - ele.vz()) < dz_cleaning_ &&
-		      reco::deltaR(ele, pfele) < dr_cleaning_
-		      );
-      }
-      if(clean_out) continue;
-      ele.addUserInt("isPF", 0);
-      ele.addUserInt("isLowPt", 1);
-      ele.addUserFloat("chargeMode", ele.gsfTrack()->chargeMode());
-      if(use_gsf_mode_for_p4_) {
-	reco::Candidate::PolarLorentzVector p4(
-					     ele.gsfTrack()->ptMode(),
-					     ele.gsfTrack()->etaMode(),
-					     ele.gsfTrack()->phiMode(),
-					     ele.mass()
-					     );
-	ele.setP4(p4);
-      }
-      alreadySavedLPT[icount] = 1;
-      out->push_back(ele);
-    }
+  // output
+  std::unique_ptr<pat::ElectronCollection> ele_out(new pat::ElectronCollection);
 
-    std::sort(
-	      out->begin(), out->end(), 
-	      [] (pat::Electron e1, pat::Electron e2) -> bool {return e1.pt() > e2.pt();}
+  
+  // -> changing order of loops ert Arabella's fix this without need for more vectors  
+ for(auto ele : *pf) {
+   //cuts
+   if (ele.pt()<ptMin_) continue;
+   if (fabs(ele.eta())>etaMax_) continue;
+   // apply conversion veto unless we want conversions
+   if (!ele.passConversionVeto()) continue;
+
+   // skip electrons inside tag's jet or from different PV
+   bool skipEle=true;
+   for(const auto & trg : *trgMuon) {
+     if(reco::deltaR(ele, trg) < drTrg_cleaning_ && drTrg_cleaning_ > 0)
+        continue;
+     if(fabs(ele.vz() - trg.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ > 0)
+        continue;
+     skipEle=false;
+     break; // one trg muon to pass is enough :)
+   }
+   // we skip evts without trg muon
```

Comment:
Given that this code is repeated exactly below, please make it an inline function, or filter the electrons after they are merged.

Reply:
Yes I agree i will try inline function but inside the same loop. I would try to avoid another filter or another loop if that's possible, again for performance issues